### PR TITLE
add crypto polyfill to allow use of Web Crypto API in workers

### PR DESCRIPTION
**Pull request by tootedom**
This pull request it to allow the use of the `Web Crypto API` in workers.
In combination with the pull (https://github.com/dollarshaveclub/cloudworker/pull/35) and (https://github.com/dollarshaveclub/cloudworker/pull/33), it would support the signing requests example at:

- https://developers.cloudflare.com/workers/recipes/signed-requests/

I've used node-webcrypto-ossl as it had support for hmac and sha256, and is in active-ish development.  If you know of another lib, then I could swap and check the e2e tests (as part of this pull) still pass.

### CODE REVIEW

File: lib/__tests__/cloudworker-e2e.test.js [hunk 1]
```diff
@@ -211,4 +211,186 @@ describe('cloudworker-e2e', async () => {
...
+    const res = await axios.get('http://localhost:8080', defaultAxiosOpts)
+    expect(res.status).toEqual(200)
+    expect(res.data).toContain('?mac=ae0342a9b989d76cfc457439f615ed72651ba7bce1994361621a003f6ade5508&expiry=1362078000000')
+    await server.close()
+    cb()
+  })
+
+  test('simple hmac verify', async (cb) => {
+    const script = `
+      addEventListener('fetch', event => {
```
Comment (hankjacobs) [inline lib/__tests__/cloudworker-e2e.test.js]: For long e2e scripts like this, please add as a fixture file instead. It helps with readability.

File: lib/__tests__/cloudworker-e2e.test.js [hunk 2]
```diff
@@ -211,4 +211,186 @@ describe('cloudworker-e2e', async () => {
...
+    expect(res.status).toEqual(200)
+    // taken from `echo -n "hello" | shasum -a 256`
+    expect(res.data).toContain('2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824')
+    await server.close()
+    cb()
+  })
+
+  test('simple hmac sign', async (cb) => {
+    const script = `
+      addEventListener('fetch', event => {
```
Comment (hankjacobs) [inline lib/__tests__/cloudworker-e2e.test.js]: For long e2e scripts like this, please add as a fixture file instead. It helps with readability.

File: lib/__tests__/cloudworker-e2e.test.js [hunk 3]
```diff
@@ -211,4 +211,186 @@ describe('cloudworker-e2e', async () => {
     await upstream.close()
     cb()
   })
+
+  test('simple sha256 encode', async (cb) => {
+    const script = `
+      addEventListener('fetch', event => {
```
Comment (hankjacobs) [inline lib/__tests__/cloudworker-e2e.test.js]: For long e2e scripts like this, please add as a fixture file instead. It helps with readability.

File: lib/__tests__/cloudworker-e2e.test.js [hunk 4]
```diff
@@ -211,4 +211,186 @@ describe('cloudworker-e2e', async () => {
...
+      async function toSHA256(message) {
+        let hash = await crypto.subtle.digest('SHA-256', byteStringToUint8Array(message))
+        return toHexString(hash)
+      }
+    `
+    const server = new Cloudworker(script).listen(8080)
+    const res = await axios.get('http://localhost:8080', defaultAxiosOpts)
+    expect(res.status).toEqual(200)
+    // taken from `echo -n "hello" | shasum -a 256`
+    expect(res.data).toContain('2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824')
```
Comment (hankjacobs) [inline lib/__tests__/cloudworker-e2e.test.js]: Let's be more specific with this check and use toEqual unless part of the response is non-deterministic.

File: lib/__tests__/cloudworker-e2e.test.js [hunk 5]
```diff
@@ -211,4 +211,186 @@ describe('cloudworker-e2e', async () => {
...
+        for (let i = 0; i < byteString.length; ++i) {
+          ui[i] = byteString.charCodeAt(i)
+        }
+        return ui
+      }
+    `
+    const server = new Cloudworker(script).listen(8080)
+    const res = await axios.get('http://localhost:8080', defaultAxiosOpts)
+    expect(res.status).toEqual(200)
+    expect(res.data).toContain('?mac=ae0342a9b989d76cfc457439f615ed72651ba7bce1994361621a003f6ade5508&expiry=1362078000000')
```
Comment (hankjacobs) [inline lib/__tests__/cloudworker-e2e.test.js]: Let's be more specific with this check and use toEqual unless part of the response is non-deterministic.

File: lib/__tests__/cloudworker-e2e.test.js [hunk 6]
```diff
@@ -211,4 +211,186 @@ describe('cloudworker-e2e', async () => {
...
+        for (let i = 0; i < byteString.length; ++i) {
+          ui[i] = byteString.charCodeAt(i)
+        }
+        return ui
+      }
+    `
+    const server = new Cloudworker(script).listen(8080)
+    const res = await axios.get('http://localhost:8080/verify/bob?mac=ae0342a9b989d76cfc457439f615ed72651ba7bce1994361621a003f6ade5508&expiry=1362078000000', defaultAxiosOpts)
+    expect(res.status).toEqual(403)
+    expect(res.data).toContain('URL expired')
```
Comment (hankjacobs) [inline lib/__tests__/cloudworker-e2e.test.js]: Let's be more specific with this check and use toEqual unless part of the response is non-deterministic.

**Comment by hankjacobs** [comment]
Also, please fix the linter errors. Additionally, it'd be great if we could include the [two scripts used in the Cloudflare Worker recipe](https://developers.cloudflare.com/workers/recipes/signed-requests/) docs in the e2e tests. I'd like to ensure every example Cloudflare has is runnable by cloudworker.

**Comment by tootedom** [comment]
Hi there,

Thanks for the review,  I'll see what I can do in the next few days.  The cloudflare examples are using the TextEncoder, so I need to crack that nut first (I'm making a bit of progress on the TextDecoder aspect).

Cheers
/dom

**Comment by hankjacobs** [comment]
@tootedom Let me know when you're ready for another review. It'd also be great to see those recipes added as e2e tests. Thanks!

**Comment by tootedom** [comment]
I've updated to merge in master (the text encoder).  
Would you like me to look at putting in the two scripts used in the Cloudflare Worker recipe (https://developers.cloudflare.com/workers/recipes/signed-requests/)?

**Comment by hankjacobs** [comment]
Yes, please.

**Comment by tootedom** [comment]
Cool, just seen your comment (which answers my question).
I'll have a look into the recipes.

Will ping when I've got something
cheers
/dom

**Comment by tootedom** [comment]
Hola,
I've added the verify and generate on (https://developers.cloudflare.com/workers/recipes/signed-requests/).

The only change I've made is to not have the "verify" fetch the generate url at the end.  All I have done is made it return `true` and `200`.  This is so I can implement the test.  

let me know.
cheers
/dom

**Comment by hankjacobs** [comment]
Excellent! Thank you! I appreciate your support. I'll review tomorrow.

**Pull request merged**